### ISSUE TITLE
Migration from 1.3.2 to 2.0.0-dev does not work

### ISSUE THREAD

**Issue by kkujawa-sumo** [SumoLogic/sumologic-kubernetes-collection#1211]
**Describe the bug**
Migration from 1.3.2 to 2.0.0-dev does not work

**Command used to install/upgrade Collection**
```
helm upgrade --install my-release sumologic/sumologic \
  --version 1.3.2 \
  -f /sumologic/vagrant/values-sumo.yaml
```
and 
```
helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml --force
```

**Configuration**
values-sumo.yaml:
```
kube-prometheus-stack:
  prometheus:
    prometheusSpec:
      externalLabels:
        cluster: microk8s
      resources:
        limits:
          cpu: 4000m
          memory: 12Gi
        requests:
          cpu: 500m
          memory: 512Mi

sumologic:
  accessId: ***
  accessKey: ***
  endpoint: ***
  cleanUpEnabled: false

```

**To Reproduce**
```
$ helm repo add sumologic https://sumologic.github.io/sumologic-kubernetes-collection
$ helm repo update
```

```
$ helm upgrade --install my-release sumologic/sumologic \
  --version 1.3.2 \
  -f /sumologic/vagrant/values-sumo.yaml
```

```
$ helm ls
NAME      	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART          	APP VERSION
my-release	default  	1       	2020-12-02 14:20:33.640994561 +0000 UTC	deployed	sumologic-1.3.2	1.3.2
```

```
$ helm repo add sumologic-dev https://sumologic.github.io/sumologic-kubernetes-collection/dev
$ helm repo update
```

```
$ ./apply.sh
```

apply.sh: 
```
#!/bin/bash

kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_probes.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_alertmanagers.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_alertmanagerconfigs.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_podmonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.43/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
```

```
$ helm get values my-release  > current_values.yaml
```

```
$ curl -LJO https://raw.githubusercontent.com/SumoLogic/sumologic-kubernetes-collection/main/deploy/helm/sumologic/upgrade-2.0.0.sh
$ chmod +x upgrade-2.0.0.sh && ./upgrade-2.0.0.sh current_values.yaml
```

```
$ helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml --force
Error: UPGRADE FAILED: failed to replace object: Service "my-release-kube-state-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-prometheus-node-exporter" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-events" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-logs" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: DaemonSet.apps "my-release-fluent-bit" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"my-release", "app.kubernetes.io/name":"fluent-bit"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-events" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-metrics" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-logs" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden
```

```
$ helm ls
NAME      	NAMESPACE	REVISION	UPDATED                                	STATUS	CHART                             	APP VERSION
my-release	default  	2       	2020-12-02 14:24:46.367215501 +0000 UTC	failed	sumologic-2.0.0-dev.0-150-g7b64cc1	2.0.0-dev.0-150-g7b64cc1
```

**Expected behavior**
Successful migration to 2.0.0-dev

**Environment:**
 - **Kubernetes version**: 
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.9", GitCommit:"94f372e501c973a7fa9eb40ec9ebd2fe7ca69848", GitTreeState:"clean", BuildDate:"2020-09-16T13:56:40Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.9", GitCommit:"94f372e501c973a7fa9eb40ec9ebd2fe7ca69848", GitTreeState:"clean", BuildDate:"2020-09-16T13:47:43Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
 - **Helm version**:
```
helm version
version.BuildInfo{Version:"v3.2.4", GitCommit:"0ad800ef43d3b826f31a5ad8dfbb4fe05d143688", GitTreeState:"clean", GoVersion:"go1.13.12"}
```

**Comment by kkujawa-sumo** [comment]
I tested with different helm versions using the same steps as above:
```
$ helm version
version.BuildInfo{Version:"v3.1.2", GitCommit:"d878d4d45863e42fd5cff6743294a11d28a9abce", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

```
$ helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml --force
Error: UPGRADE FAILED: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: namespace: , name: my-release-kube-state-metrics, existing_kind: rbac.authorization.k8s.io/v1, Kind=ClusterRole, new_kind: rbac.authorization.k8s.io/v1, Kind=ClusterRole
```

and

```
$ helm version
WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /var/snap/microk8s/current/credentials/kubelet.config
version.BuildInfo{Version:"v3.4.0", GitCommit:"7090a89efc8a18f3d8178bf47d2462450349a004", GitTreeState:"clean", GoVersion:"go1.14.10"}
```

```
$ helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml --force
WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /var/snap/microk8s/current/credentials/kubelet.config
Error: UPGRADE FAILED: failed to replace object: Service "my-release-kube-state-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-prometheus-node-exporter" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-events" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-metrics" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: Service "my-release-sumologic-fluentd-logs" is invalid: spec.clusterIP: Invalid value: "": field is immutable && failed to replace object: DaemonSet.apps "my-release-fluent-bit" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"my-release", "app.kubernetes.io/name":"fluent-bit"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-events" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-metrics" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden && failed to replace object: StatefulSet.apps "my-release-sumologic-fluentd-logs" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden
```

**Comment by kkujawa-sumo** [comment]
These errors:
```
 cannot patch "my-release-sumologic-fluentd-events" with kind StatefulSet: 
 StatefulSet.apps "my-release-sumologic-fluentd-events" is invalid: spec: 
 Forbidden: updates to statefulset spec for fields other than 
 'replicas', 'template', and 'updateStrategy' are forbidden &&

 cannot patch "my-release-sumologic-fluentd-metrics" with kind StatefulSet: 
 StatefulSet.apps "my-release-sumologic-fluentd-metrics" is invalid: 
 spec: Forbidden: updates to statefulset spec for fields other than 
 'replicas', 'template', and 'updateStrategy' are forbidden &&

 cannot patch "my-release-sumologic-fluentd-logs" with kind StatefulSet: 
 StatefulSet.apps "my-release-sumologic-fluentd-logs" is invalid: spec: 
 Forbidden: updates to statefulset spec for fields other than 
 'replicas', 'template', and 'updateStrategy' are forbidden
```
disappear when I add `--set fluentd.persistence.enabled=false` for `helm upgrade` command

**Comment by kkujawa-sumo** [comment]
For this error:
```
Error: UPGRADE FAILED: cannot patch "my-release-fluent-bit" with kind DaemonSet: 
DaemonSet.apps "my-release-fluent-bit" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"my-release", 
"app.kubernetes.io/name":"fluent-bit"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```
helped these changes in helm chart:

```
{{/*
Selector labels
*/}}
{{- define "fluent-bit.selectorLabels" -}}
app: {{ include "fluent-bit.name" . }}
release: {{ .Release.Name }}
{{- end -}}
```

**Comment by sumo-drosiek** [comment]
> For this error:
> 
> ```
> Error: UPGRADE FAILED: cannot patch "my-release-fluent-bit" with kind DaemonSet: 
> DaemonSet.apps "my-release-fluent-bit" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"my-release", 
> "app.kubernetes.io/name":"fluent-bit"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
> ```
> 
> helped these changes in helm chart:
> 
> ```
> {{/*
> Selector labels
> */}}
> {{- define "fluent-bit.selectorLabels" -}}
> app: {{ include "fluent-bit.name" . }}
> release: {{ .Release.Name }}
> {{- end -}}
> ```

Which helm chart and which line do you refer to?

**Comment by kkujawa-sumo** [comment]
I made this changes in unpacked  helm chart in version 2.0.0-dev.0-150-g7b64cc1 in `sumologic/charts/fluent-bit/templates/_helpers.tpl`

**Comment by kkujawa-sumo** [comment]
> For this error:
> 
> ```
> Error: UPGRADE FAILED: cannot patch "my-release-fluent-bit" with kind DaemonSet: 
> DaemonSet.apps "my-release-fluent-bit" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"my-release", 
> "app.kubernetes.io/name":"fluent-bit"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
> ```
> 
> helped these changes in helm chart:
> 
> ```
> {{/*
> Selector labels
> */}}
> {{- define "fluent-bit.selectorLabels" -}}
> app: {{ include "fluent-bit.name" . }}
> release: {{ .Release.Name }}
> {{- end -}}
> ```

spec.selector is immutable since k8s v1.16, see [this](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/)

 sumologic v1.3.2 uses fluent-bit with following definition of DaemonSet:
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: my-release-fluent-bit
  labels:
    app: fluent-bit
    chart: fluent-bit-2.10.1
    heritage: Helm
    release: my-release
spec:
  selector:
    matchLabels:
      app: fluent-bit
      release: my-release
...
```

sumologic v2.0.0-dev.0-150-g7b64cc1 uses fluent-bit with following definition of DaemonSet:

```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: my-release-fluent-bit
  labels:
    helm.sh/chart: fluent-bit-0.7.10
    app.kubernetes.io/name: fluent-bit
    app.kubernetes.io/instance: my-release
    app.kubernetes.io/version: "1.6.6"
    app.kubernetes.io/managed-by: Helm
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: fluent-bit
      app.kubernetes.io/instance: my-release
...
```

**Comment by kkujawa-sumo** [comment]
just for reference another error appear for helm v3.1.0 and helm v.3.3.0

```
$ helm version
version.BuildInfo{Version:"v3.1.0", GitCommit:"b29d20baf09943e134c2fa5e1e1cab3bf93315fa", GitTreeState:"clean", GoVersion:"go1.13.7"}
$helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml
Error: UPGRADE FAILED: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: namespace: , name: my-release-kube-state-metrics, existing_kind: rbac.authorization.k8s.io/v1, Kind=ClusterRole, new_kind: rbac.authorization.k8s.io/v1, Kind=ClusterRole
```

```
$ helm version
version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
$ helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml 
Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with update: ClusterRole "my-release-kube-state-metrics" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "my-release"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "default"
```

**Comment by sumo-drosiek** [comment]
> ```
> $ helm version
> version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
> $ helm upgrade my-release sumologic-dev/sumologic --version=2.0.0-dev.0-150-g7b64cc1 -f new_values.yaml 
> Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with update: ClusterRole "my-release-kube-state-metrics" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "my-release"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "default"
> ```

Is it from vagrant?

**Comment by kkujawa-sumo** [comment]
I deployed this on vagrant VM machine from [release-v1.3](https://github.com/SumoLogic/sumologic-kubernetes-collection/tree/release-v1.3).

**Comment by kkujawa-sumo** [comment]
Issue which impacts migration to 2.0.0 using helm3: https://github.com/helm/helm/issues/7082
See also explanation of differences between helm2 and helm3: https://github.com/helm/helm/issues/7082#issuecomment-559558318

**Comment by kkujawa-sumo** [comment]
Fixed by #1242
